### PR TITLE
AP-4617: Use puffing-billy to stub feature test javascript API calls

### PR DESCRIPTION
**Pull request by jsugarman**
## What
Implement puffing-billy, an intercepting proxy server for JS requests, to prevent actual LFA (and other) API calls during  feature test runs.

The aim is not to fix all of them but introduce pattern that can be used going forward.

[Link to story](https://dsdmoj.atlassian.net/browse/AP-4617)

So test suite can pass when external APIs are unavailable. LFA is not available between 22:00 and 07:00, for example.

<details>
<summary> Ensure all http client requests are intercepted by VCR</summary>
<a href="https://www.rubydoc.info/gems/vcr/VCR%2FConfiguration:hook_into">VCR::Configuration#hook_into</a>
<a href="https://github.com/vcr/vcr/blob/master/features/configuration/hook_into.feature">hook_into documentation</a>

The following http client libraries are used and therefore we need VCR
to intercept them.

| API              | Library   |
| ---------------- | --------- |
| Address lookup   | faraday   |
| Bank holidays    | Net::HTTP |
| LFA              | faraday   |
</details>

<details>
<summary> Ensure JS script initiated requests are intercepted</summary>
The test suite should pass locally when wifi is disabled
</details>

## TODO

- [x] Get screenshot helper working with billy chrome driver
- [x] Get worst offending JS reliant features stubbed - all should pass with wifi off
- [x] Document the setup, creation of and/or recording of stubs

## Offending tests

- [x] features/providers/applicant_details.feature:3 *1
- [x] features/providers/applicant_details.feature:132 
- [x] features/providers/applicant_details.feature:191 
- [x] features/providers/applicant_details.feature:267 
- [x] features/providers/applicant_details.feature:320 
- [x] features/providers/applicant_details.feature:380 *1
- [x] features/providers/regressions/scope_limitations.feature:3 *3
- [x] features/providers/merits_task_list_opponent.feature:24 *3
- [x] features/providers/check_provider_answers.feature:83 *3
- [x] features/providers/non_means_tested_journey/under_18_application_journey.feature:4 *3
- [x] features/providers/cost_override.feature:3  *3
- [x] features/providers/search_proceedings.feature:4 *3
- [x] features/providers/search_proceedings.feature:25 *3
- [x] features/providers/non_means_tested_journey/without_delegated_functions.feature:4 *2
- [x] features/providers/non_means_tested_journey/with_delegated_functions.feature:4 *2
- [x] features/providers/search_opponent_organisations.feature:15
- [x] features/providers/search_opponent_organisations.feature:21
- [x] features/providers/search_opponent_organisations.feature:27
- [x] features/providers/search_opponent_organisations.feature:34
- [x] features/providers/search_opponent_organisations.feature:47
- [x] features/providers/search_opponent_organisations.feature:59
- [x] features/providers/pathways_from_check_your_answers.feature:79 *2
- [x] features/providers/no_nino_flow.feature:4 *2
- [x] features/providers/regressions/back_button_on_proceedings.feature:3 
- [x] features/providers/applicant_under_16_blocked.feature:4 *2

*\*1 bank holiday query still called (by worker job) [AP-4686](https://dsdmoj.atlassian.net/browse/AP-4686)*
*\*2 see [AP-4685](https://dsdmoj.atlassian.net/browse/AP-4685)*
*\*3 see [AP-4684](https://dsdmoj.atlassian.net/browse/AP-4684)*
## Checklist

Before you ask people to review this PR:

- Tests and rubocop should be passing: `bundle exec rake`
- Github should not be reporting conflicts; you should have recently run `git rebase main`.
- There should be no unnecessary whitespace changes. These make diffs harder to read and conflicts more likely.
- The PR description should say what you changed and why, with a link to the JIRA story.
- You should have looked at the diff against main and ensured that nothing unexpected is included in your changes.
- You should have checked that the commit messages say why the change was made.

[AP-4686]: https://dsdmoj.atlassian.net/browse/AP-4686?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/puffing_billy_recording.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Puffing-billy stub creating and request cache
+
+Some feature tests excercise application logic that includes aysnchronous javascript
+calls to external services (APIs). In particular searching for proceedings and organisations is achieved this way. Because VCR does intercept browser calls puffing-billy's "rewriting web proxy" is used to achieve their stubbing. VCR and puffing-billy can be used together to achieve stubbing of any calls to external services.
```
Comment (agoldstone93) [inline docs/puffing_billy_recording.md:4]: Is there a 'not' missing in this line?

File: features/providers/search_opponent_organisations.feature [hunk 2]
```diff
@@ -11,26 +11,26 @@ Background: User is on existing opponent organisation search page
   And I click 'Save and continue'
   Then I should be on a page with title "Opponent"
 
-@javascript
+@javascript @billy
 Scenario: No results returned is seen on screen when non-existant organisation search entered
-  When I search for organisation "cakes"
+  When I search for organisation "cake"
```
Comment (agoldstone93) [inline features/providers/search_opponent_organisations.feature:16]: Funny change. This made me laugh. Does this actually make a difference???

**Review by agoldstone93 — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -201,6 +201,12 @@ VCR is used to record interactions with external services and play back these st
 
 see [VCR Recording](docs/vcr_recording.md) for setup and recommended approaches to rerecording of cassettes.
 
+#### Puffing-billy stubs (and request cache)
+
+[Puffing-billy](https://github.com/oesmith/puffing-billy) is used to stub (or record) external service calls made from javascript.
+
+see [Puffing-blly recording](docs/puffing_billy_recording.md) for setup and recommended approaches to creating new stubs recording requests in persisted cache.
```
Comment (jsugarman) [inline README.md:208]:
```suggestion
see [Puffing-billy recording](docs/puffing_billy_recording.md) for setup and recommended approaches to creating new stubs recording requests in persisted cache.
```

File: docs/puffing_billy_recording.md (hunk 1 above)
Comment (jsugarman) [inline docs/puffing_billy_recording.md:4]:
```suggestion
calls to external services (APIs). In particular searching for proceedings and organisations is achieved this way. VCR does not intercept browser calls therefore puffing-billy is used. This provides a "rewriting web proxy" to achieve the interception and stubbing. VCR and puffing-billy can be used together to achieve stubbing of any calls to external services from backend and frontend respectively.
```

File: features/providers/search_opponent_organisations.feature (hunk 2 above)
Comment (jsugarman) [reply features/providers/search_opponent_organisations.feature:16]: 😄 no, i changed it to something else and then changed it back to `cake`.  In reality anything which is not one of the stubbed search terms will return the same stubbed response (`{ success: false, data: [] }`)

Comment (agoldstone93) [reply features/providers/search_opponent_organisations.feature:16]: I assumed not!

**Review by agoldstone93 — APPROVED** [review]

**Review by kmahern — APPROVED** [review]
LGTM

**Pull request merged**